### PR TITLE
Add Coxeter Groups and Root Systems

### PR DESCRIPTION
**Pull request by felix-roehrich**
I am currently working on implementing https://github.com/ulthiel/CoxeterGroups.jl (see https://github.com/oscar-system/Oscar.jl/pull/2217). To have compatibility with root systems and since many efficient algorithms for Weyl groups rely on geometric properties, I have also written a first draft for root systems in OSCAR.
Currently this is still under heavy development, especially distributing logic between root systems and Weyl groups.
At the moment Cartan matrices are implemented as ZZMatrix, but it might make sense to implement it as a sparse matrix instead.
Documentation and tests are also missing.

For simplicity everything lives in experimental/LieAlgebras but everything from Coxeter groups that does not depend on root systems should probably moved to the Groups folder at some point.

### CODE REVIEW

**Comment by simonbrandhorst** [comment]
Note that we have root lattices in the sense of quadratic forms coming from Hecke:
```
julia> root_lattice(:A,4)
Integer lattice of rank 4 and degree 4
with gram matrix
[ 2   -1    0    0]
[-1    2   -1    0]
[ 0   -1    2   -1]
[ 0    0   -1    2]
```
Perhaps they can be of use to you.
I did not have root systems in mind when I implemented them though. But for some application I needed a highest root.
```
julia> highest_root(:A, 4)
[1   1   1   1]
```

**Comment by felix-roehrich** [comment]
Thanks, I'll have a look.

File: experimental/LieAlgebras/src/AbstractRootSystem.jl
Comment (fingolfin) [inline experimental/LieAlgebras/src/AbstractRootSystem.jl]: This seems to suggest that only irreducible root systems are supported?

While that is of course fine for now, on the long run it'd be nice to support more general ones.

File: experimental/LieAlgebras/src/AbstractRootSystem.jl
Comment (fingolfin) [inline experimental/LieAlgebras/src/AbstractRootSystem.jl]: I think this is an unfortunate choice of names, given the meaning of `abstract` in the Julia language.

What is the idea behind having an abstract type `RootSystem`, and a non-abstract type `AbstractRootSystem`, anyway?

File: experimental/LieAlgebras/src/AbstractRootSystem.jl
Comment (fingolfin) [inline experimental/LieAlgebras/src/AbstractRootSystem.jl]:
```suggestion
  error("TODO, not implemented")
end
```

**Comment by fingolfin** [comment]
Thank you for this, I'll leave some comments later. There are many other people which are working on things in algebraic Lie theory which may have an interest in this, e.g. @lgoettgens and @voggesbe. We don't need to have this perfectly in sync but it would be good if eventually we can share types for things like root systems between Lie algebras, Coxeter/Weyl groups and other applications, so it's good to stay in touch about these things

**Comment by fingolfin** [comment]
One further comment: on the long run, I want to work with infinite Coxeter groups, too, and also work on code for Kac-Moody groups and algebras. For these (but also already for Lie algebras and algebraic groups) it may be useful to also have a notion of "root datum" (like Magma also has, see https://magma.maths.usyd.edu.au/magma/handbook/text/1143). I am merely saying this to set a vision for the future; this is not a request for you to change anything here right now.

File: experimental/LieAlgebras/src/CartanMatrix.jl
Comment (fingolfin) [inline experimental/LieAlgebras/src/CartanMatrix.jl:43]: Just as a side remark (not a change request): Actually there is also $E_9$, $E_{10}$ - the resulting Weyl/Coxeter groups, Lie algebras just are not finite dimensional anymore. I have particular interest in these, so ultimately I'd want this to work more generally.

File: experimental/LieAlgebras/src/CoxeterGroup.jl
Comment (fingolfin) [inline experimental/LieAlgebras/src/CoxeterGroup.jl]: please always end files with a newline (you can see in the GitHub file view of your PR that it complains about this here)

File: experimental/LieAlgebras/src/RootSystem.jl
Comment (fingolfin) [inline experimental/LieAlgebras/src/RootSystem.jl]:
```suggestion
abstract type RootSystem end

```

File: experimental/LieAlgebras/src/WeylGroup.jl
Comment (fingolfin) [inline experimental/LieAlgebras/src/WeylGroup.jl]: Maybe you mean this?
```suggestion
#   Weyl Groups
```

File: experimental/LieAlgebras/src/WeylGroup.jl
Comment (fingolfin) [inline experimental/LieAlgebras/src/WeylGroup.jl:10]: you probably will want this to be attribute storing on the long run...
```suggestion
@attributes mutable struct WeylGroup <: CoxeterGroup
```

I think it would be useful to have a comment (or better: docstring) that briefly explains in how far you consider "Weyl group" as a special case of Coxeter group (i.e., what makes it different from "other" kinds of of Coxeter groups)

**Comment by lgoettgens** [comment]
> Thank you for this, I'll leave some comments later. There are many other people which are working on things in algebraic Lie theory which may have an interest in this, e.g. @lgoettgens and @voggesbe. We don't need to have this perfectly in sync but it would be good if eventually we can share types for things like root systems between Lie algebras, Coxeter/Weyl groups and other applications, so it's good to stay in touch about these things

@felix-roehrich has his office two doors down from mine. And we talk regularly about this, and how it could connect to the other Lie algebra stuff in Oscar.

**Review by fingolfin — APPROVED** [review]

**Comment by fingolfin** [comment]
@felix-roehrich @lgoettgens are you also in contact with @ulthiel about this?

File: experimental/LieAlgebras/src/AbstractRootSystem.jl
Comment (lgoettgens) [reply experimental/LieAlgebras/src/AbstractRootSystem.jl]: I think this is similar to (currently) Lie algebras in Oscar. `LieAlgebra` is the abstract supertype, while `AbstractLieAlgebra` is a concrete type for Lie algebras given only by structure constants (named after Humphreys: "Introduction to Lie algebras and representation theory", Chap. 1.4). Of course, this can be changed as well, but there wasn't a better name suggested yet.

**Comment by felix-roehrich** [comment]
I am thankful for your comments but please wait until I mark the PR as ready for review. Most of the comments you just made are already outdated from my recent work and this will just create more work for both of us. I will mark your recent comments as resolved once I push my changes.

File: experimental/LieAlgebras/src/CartanMatrix.jl
Comment (felix-roehrich) [reply experimental/LieAlgebras/src/CartanMatrix.jl:43]: If you have a reference to common generalized Cartan matrices, I will implement them. For now I have implemented the classical case, but when offering generalized Cartan matrices, there should be some rules on what is offered by the helper.

**Comment by felix-roehrich** [comment]
Tests and sanity checks still missing. I will add those if no more (big) changes are requested/necessary.

File: experimental/LieAlgebras/src/AbstractRootSystem.jl
Comment (felix-roehrich) [reply experimental/LieAlgebras/src/AbstractRootSystem.jl]: General root systems are supported, this was used as a helper for printing. For now I have removed this. In general we can only print the Cartan matrix defining the root system, since detecting the Dynkin type of the Cartan matrix is not easy.
I am not sure what a nice `show` for a root system (and Weyl group) would be; feel free to make suggestions.

**Comment by lgoettgens** [comment]
I'll hope to find some time the next days to review this. And it would be great if @fingolfin could have a look as well.

File: experimental/LieAlgebras/src/WeylGroup.jl
Comment (felix-roehrich) [reply experimental/LieAlgebras/src/WeylGroup.jl:10]: `WeylGroup` is the implemtation of `CoxGrpMin` in https://github.com/ulthiel/CoxeterGroups.jl. Since this PR is quite large already, in a follow up PR I want to add `CoxGrpRec` renamed to `CoxeterGroupGeneric` (?) and offer a generic Coxeter group constructor from a Coxeter matrix. In my opinion there should then be references for specials cases like Weyl Groups (with specialized algorithms). In https://github.com/ulthiel/CoxeterGroups.jl/issues/1 `CoxeterGroups{T}` was suggested. I am open for suggestions also from @ulthiel.
Also note that an abstract type is necessary to break cyclic dependency between root systems and Weyl groups.

(Maybe I also misunderstood your comment.)

**Comment by ulthiel** [comment]
I'll test it. But before I forget: please make sure that the following people are mentioned as contributors (as well): Cameron Braunstein (@CameronBraunstein), Joel Gibson (University of Sydney, @joelgibson), Tom Schmit (@schto223). As with JuLie, if this gets merged then I'll archive CoxeterGroups.jl.

On the type design/hierarchy the experts should comment (@fingolfin). I still don't know what's the way to go. (It would be good to have Coxeter groups finally, though, in any way.)

File: experimental/LieAlgebras/src/CartanMatrix.jl
Comment (ulthiel) [reply experimental/LieAlgebras/src/CartanMatrix.jl:43]: The standard reference for Cartan matrices (generalized) should be Kac's book [Infinite Dimensional Lie Algebras](https://link.springer.com/book/10.1007/978-1-4757-1382-4) I would say.

File: experimental/LieAlgebras/src/WeylGroup.jl
Comment (ulthiel) [reply experimental/LieAlgebras/src/WeylGroup.jl:10]: Whenever I read "Weyl group" I think "finite Coxeter group". Maybe I am too much used to the algebraic groups setting. Kac uses Weyl group more generally as well. I don't know what's best. But note that CoxGrpMin refers to an _algorithm_ (namely using minimal roots). I somehow like(d) to make this clear. But again, I don't know what's best.

Comment (ulthiel) [reply experimental/LieAlgebras/src/WeylGroup.jl:10]: 2 days later I really think there should be separate types for the separate algorithms for computing in (some classes of) Coxeter groups and the types should also be called like that, as originally in CoxeterGroups.jl. So, I would prefer CoxGrpMin to WeylGroup. Also, there is CoxGrpSym in CoxeterGroups.jl which is extremely useful for type-A computations. This should go in here as well. There may be CoxGrpPerm for a permutation group algorithm etc. The generic function coxeter_group(...) should select the type that's most appropriate for the given type. @fingolfin, what do you think?

Comment (fingolfin) [reply experimental/LieAlgebras/src/WeylGroup.jl:10]: I will look at this next week.

**Comment by lgoettgens** [comment]
> I'll hope to find some time the next days to review this. And it would be great if @fingolfin could have a look as well.

I unfortunately didn't get to this yet. And as I am on holidays until the 8th, I'll only manage to review this the week after.

**Comment by fingolfin** [comment]
Both PR #2572 and PR #2689 add root systems. How shall we resolve this? @felix-roehrich @voggesbe @lgoettgens ?

**Comment by lgoettgens** [comment]
> Both PR #2572 and PR #2689 add root systems. How shall we resolve this? @felix-roehrich @voggesbe @lgoettgens ?

According to @felix-roehrich there should be no conflict in merging both after review. Afterwards, we can try to unify these. My dream would be to make the GAP-backed root systems a concrete subtype of the abstract root system type of this PR.

File: experimental/LieAlgebras/src/AbstractRootSystem.jl
Comment (lgoettgens) [inline experimental/LieAlgebras/src/AbstractRootSystem.jl]: `$v` does not exist here. 

Below, you only use `w` instead of `$w`. It would be great to have that consistently

**Comment by fingolfin** [comment]
> My dream would be to make the GAP-backed root systems a concrete subtype of the abstract root system type of this PR.

I don't quite share that -- it sounds as if we would have multiple root system implementations that overlap in functionality but are still different. That's something I'd rather avoid. But perhaps I misunderstand?

**Comment by lgoettgens** [comment]
> > My dream would be to make the GAP-backed root systems a concrete subtype of the abstract root system type of this PR.
> 
> I don't quite share that -- it sounds as if we would have multiple root system implementations that overlap in functionality but are still different. That's something I'd rather avoid. But perhaps I misunderstand?

Let's get #2572 merged first. Afterwards, I can work with @felix-roehrich on making them compatible / merging them /... let's see how it evolves.
I find it quite hard to reason about it currently with everything living in two different PRs. With one in master and the other one in a PR, this will be a lot simpler.

And after all, this is still in experimental.

**Comment by felix-roehrich** [comment]
> Let's get #2572 merged first. Afterwards, I can work with @felix-roehrich on making them compatible / merging them /... let's see how it evolves.

If I didn't miss anything #2572 only uses GAP only to get basic root system information (like number of roots, Cartan matrix). So this can easily be replaced with the implementation in this PR.

**Review by fingolfin — APPROVED** [review]
@lgoettgens @voggesbe sorry for not having time to review these two PRs properly. I still want to do that, but I also don't want to be a bottleneck. So I will just approve #2572 and #2689 for now and leave it to you to figure out in which order to merge them. I will nevertheless try to eventually review them, but it'll be fine to heed my comments only post-merge (again, depending on how you can best deal with it).

The point here is that indeed it is difficult to develop them in parallel, so better to merge them and then work on converging, even if it may mean  throwing out some stuff again or redoing some. That's fine. There will be more changes in the future anyway as we start to use root system (and root data etc.) in more places (e.g. I'd like to eventually be able to construct algebraic groups via root data... or support Kac-Moody algebras... and hyperbolic Coxeter groups... and more)

File: experimental/LieAlgebras/src/AbstractRootSystem.jl
Comment (fingolfin) [inline experimental/LieAlgebras/src/AbstractRootSystem.jl]:
```suggestion
  @req w.root_system === w2.root_system "$w and $w2 must belong to the same root lattice"
```

**Review by lgoettgens — DISMISSED** [review]
I would like to wait for https://github.com/oscar-system/Oscar.jl/pull/2572 to be merged first before this.

File: experimental/LieAlgebras/src/WeylGroup.jl
Comment (ulthiel) [reply experimental/LieAlgebras/src/WeylGroup.jl:10]: Could this be looked at before merging?

File: experimental/LieAlgebras/test/WeylGroup-test.jl
Comment (lgoettgens) [inline experimental/LieAlgebras/test/WeylGroup-test.jl]:
```suggestion
end

```
Empty line at end of file

File: experimental/LieAlgebras/test/RootSystem-test.jl
Comment (lgoettgens) [inline experimental/LieAlgebras/test/RootSystem-test.jl]:
```suggestion
end

```

File: experimental/LieAlgebras/test/runtests.jl
Comment (lgoettgens) [inline experimental/LieAlgebras/test/runtests.jl]: Please revert this

Comment (felix-roehrich) [reply experimental/LieAlgebras/test/runtests.jl]: Most files had failing tests because of the broken `lie_algebra` constructor from our changes today. So I had to comment it out. I removed the comment blocks, but you will still need to fix the tests.

**Comment by lgoettgens** [comment]
This is now only missing a fix for the `AbstractLieAlgebra` constructor. I will work on this this afternoon.

**Review by lgoettgens — APPROVED** [review]
This is now good to go from my POV. Further functionalities, tests, and docs will be added by @felix-roehrich in a future PR.

Can somebody with merge rights (maybe @thofma) enable auto-squash-merge on this?

**Comment by lgoettgens** [comment]
@fingolfin please let us know if you have any urgent changes to be made in this PR. If not, please merge this and you can write comments afterward that get addressed in a follow-up.

**Comment by ulthiel** [comment]
> @fingolfin please let us know if you have any urgent changes to be made in this PR. If not, please merge this and you can write comments afterward that get addressed in a follow-up.

I made several comments/questions above on types and naming. There was no reply.

**Comment by lgoettgens** [comment]
> > @fingolfin please let us know if you have any urgent changes to be made in this PR. If not, please merge this and you can write comments afterward that get addressed in a follow-up.
> 
> I made several comments/questions above on types and naming. There was no reply.

Ah, sorry, I somehow missed them while looking through this PR. I'll give an answer here.

We decided to (for the moment at least; this can of course be renamed after a discussion) keep the name WeylGroup for the following reasons:
- OSCAR type names should convey mathematical meaning. Your suggestion CoxGrpMin is a weird combination of abbreviations (that are not to be used due to the OSCAR style guide, unless the exemptions specified there). Every mathematician not knowing about the specific internals of the implementation will be confused of the name. And if I understand the OSCAR philosophy right, we don't want users to think about such internals.
- In the current only use case (root systems of Lie algebras), the terminology of a Weyl group of that root system is most commonly used.
- Some explanation about the concrete algorithm used will be added to the docs. The same holds for the use of the word "Weyl group", which is used in a very general manner like Kac. 

Other subtypes of CoxeterGroup will get implemented by @felix-roehrich add a later point in time. 
Your suggestion to have a `coxeter_group` function that chooses the implementation based on the input will, with very high probability, be type instable. So I would refrain from using this in any library code. For user convenience, we can talk about it once there are multiple types to choose from. 

All of this is of course not set in stone and can be changed in a future PR. 
But since this PR is open for already more than 2 months, and I would like to use parts of it in #2936, I would like to get this going.

@felix-roehrich if I forgot something, please add.

@fingolfin already approved this in a very similar state. Could you please give your opinion to this?

**Comment by ulthiel** [comment]
Alright, good, it seems you thought about this, that's all I wanted. I want to protest though against calling my idea a "weird combination" :-D (joking)!

File: experimental/LieAlgebras/src/AbstractLieAlgebra.jl
Comment (fingolfin) [inline experimental/LieAlgebras/src/AbstractLieAlgebra.jl:104]: You say "the" Chevalley basis, "the" Cartan subalgebra. These are not in general unique things, so what is meant? IMHO this should either be changed to "a" in both cases, or somewhere should be an explanation in how far something "unique" is returned that warrants the use "the"

File: experimental/LieAlgebras/src/AbstractLieAlgebra.jl
Comment (fingolfin) [inline experimental/LieAlgebras/src/AbstractLieAlgebra.jl:109]: Wouldn't `root_system(L)` below already throw a similar error? Removing this check here then would have the advantage that this method would start working once someone implements runtime computation of root systems...

File: experimental/LieAlgebras/src/AbstractLieAlgebra.jl
Comment (fingolfin) [inline experimental/LieAlgebras/src/AbstractLieAlgebra.jl:320]: Docstring doesn't match changed signature
```suggestion
    lie_algebra(R::Field, S::Symbol, n::Int; cached::Bool) -> AbstractLieAlgebra{elem_type(R)}

Construct the simple Lie algebra over the ring `R` with Dynkin type given by `S` and `n`.
```

File: experimental/LieAlgebras/src/CartanMatrix.jl
Comment (fingolfin) [inline experimental/LieAlgebras/src/CartanMatrix.jl:2]:
```suggestion
```

File: experimental/LieAlgebras/src/CartanMatrix.jl
Comment (fingolfin) [inline experimental/LieAlgebras/src/CartanMatrix.jl:109]:
```suggestion
Test if `mat` is a generalized Cartan matrix. The keyword argument
`generalized` can be set to `false` to restrict this to only genuine
Cartan matrices.
```

File: experimental/LieAlgebras/src/CartanMatrix.jl
Comment (fingolfin) [inline experimental/LieAlgebras/src/CartanMatrix.jl:82]: No tests cover this function :-(

File: experimental/LieAlgebras/src/CartanMatrix.jl
Comment (fingolfin) [inline experimental/LieAlgebras/src/CartanMatrix.jl:107]:
```suggestion
    is_cartan_matrix(mat::ZZMatrix; generalized::Bool=true) -> Bool
```

File: experimental/LieAlgebras/src/RootSystem.jl
Comment (fingolfin) [inline experimental/LieAlgebras/src/RootSystem.jl:2]:
```suggestion
```

File: experimental/LieAlgebras/src/CoxeterGroup.jl
Comment (fingolfin) [inline experimental/LieAlgebras/src/CoxeterGroup.jl:3]: Meta-remark: maybe at some point in the future rename the `LieAlgebras` directory to `LieTheory` or `AlgebraicLieTheory` (of course only after all big PRs have been merged, i.e., this is far from urgent)

File: experimental/LieAlgebras/src/RootSystem.jl
Comment (fingolfin) [inline experimental/LieAlgebras/src/RootSystem.jl:11]: Couldn't this at least be, say, 
```suggestion
  positive_roots::Vector #::Vector{RootSpaceElem} (cyclic reference)
  weyl_group::CoxeterGroup     #::WeylGroup (cyclic reference)
```
The entries could be made concrete via type parameters. If it ever matters (right now I guess it doesn't).

File: experimental/LieAlgebras/src/CoxeterGroup.jl
Comment (fingolfin) [inline experimental/LieAlgebras/src/CoxeterGroup.jl:1]: Why doesn't this subtype `Group`?

File: experimental/LieAlgebras/src/RootSystem.jl
Comment (fingolfin) [inline experimental/LieAlgebras/src/RootSystem.jl:113]: This suffers from type instability, can be solved via
```suggestion
  return length(positive_roots(R))
```
Same for most other places accessing `R.positive_roots` instead of `positive_roots(R)`

File: experimental/LieAlgebras/src/RootSystem.jl
Comment (fingolfin) [inline experimental/LieAlgebras/src/RootSystem.jl:129]: Better:
```suggestion
  return positive_roots(R)[i]
```

File: experimental/LieAlgebras/src/WeylGroup.jl
Comment (fingolfin) [reply experimental/LieAlgebras/src/WeylGroup.jl:10]: To me Weyl group applies to infinite groups as well, in Kac Moody theory at least we consistently use that.

Having different types for different algorithms makes some sense, but there are also many drawbacks. But I don't think it's necessary to hold up this PR about it, as this is purely for the *experimental* branch.

If we want to make progress on this, I don't think discussions on PRs are the way forward. It just doesn't work that well. We need to sit together in a room with enough time (possible organize a little workshop) and hash it out... but in the end, actual code trumps all. If something is implemented and works, great; if it doesn't work, we can iterate on it.

I.e., "use case drive design" -- which is why I said in a PR comment elsewhere that I don't want to hold up this PR right now as I just don't have time to work on it or one code using it, and if it works for you guys right now, great. I already *know* I will need changes for my use cases (e.g. infinite root systems should not store explicit vectors of positive roots; and I need a `RootDatum` type; etc.) but all of that can (and IMHO: should) be added *later*.

**Review by fingolfin — APPROVED** [review]
Fine by me. 

I left a few comments on type stability, outdated docstrings etc. 

As I understood @lgoettgens on Friday he'd prefer to apply the corresponding fixes post-merge. Fine by me, too.

File: experimental/LieAlgebras/src/CoxeterGroup.jl
Comment (felix-roehrich) [reply experimental/LieAlgebras/src/CoxeterGroup.jl:1]: GroupsCore requires `order` to be implemented, but since a `WeylGroup` can be constructed from an arbitrary Cartan matrix, we would have to detect the type to return the order. Thus, I left this for later.

**Pull request merged**
